enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration will be required for all cloud providers.
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+A: CCM config is backward compatible with the in-tree cloud provider, therefore we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is the preferred option. It graduates in GA in 1.19 (1.20).
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and it only provides metrics and readiness/health endpoint.
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+A: KCM and CCM perform independently after the initial migration process. KCM will be unaffected by CCM being down after the initial migration. The CCM cluster operator will be responsible for transitioning to Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
+This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.
+
+Q: Does every node need a CCM?
+A: No. The cluster only needs one active `CCM` at any time. A `Deployment` will manage the `CCM` pod and will have 2 replicas which will use leader election to nominate an active leader. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)
+
+Q: How metrics are affected by the CCM migration?
+A: Currently the `CCM` implementation is using the same set of metrics from the core repository, all existing metrics will still be available post migration. The `CCMO` will be required to create a `ServiceMonitor` to instruct prometheus to collect metrics from the `CCM` pod.
+
```

Comment:
Other questions which not necessarily needs answers right away are:
1. Do we plan to run multiple CCMs for different clouds in parallel?

Reply:
Yes. This is what GCP does in their solutions according to what I heard. It is possible to do that, only correct naming and leader separation is needed.

Reply:
This question is answered in the non-goals section, we do not plan to support multiple clouds from a single cluster, so no multiple CCMs

Reply:
Agree with Joel. No, not until the need for this feature is communicated. It is possible

Reply:
Right, apologies I missed it.